database/mysql/mysql.go

```diff
@@ -117,7 +113,49 @@ func extractCustomQueryParams(c *mysql.Config) (map[string]string, error) {
 	return customQueryParams, nil
 }
 
+// extractDSNParam extracts the value of a param specified in the
+// DSN string.
+func extractDSNParam(queryString, param string) (string, error) {
+	for _, v := range strings.Split(queryString, "&") {
+		p := strings.SplitN(v, "=", 2)
+		if len(p) != 2 {
+			continue
+		}
+		if p[0] == param {
+			return nurl.QueryUnescape(p[1])
+		}
+	}
+	return "", nil
+}
+
+// ensureRegisteredTLSConfig registers a config with the supplied tls name, if not using a default config.
+// If the user is supplying a custom tls configuration, i.e. needing to supply certs using the `x-tls-*`
+// query params in the DSN, we pre-register an empty config with that name,. This avoids the errors
+// occurring when making the call to mysql.ParseDSN() which tries to lookup the tls config and throws
+// an error if it doesn't exist.
+func ensureRegisteredTLSConfig(ctls string) error {
+	switch ctls {
+	case "true", "false", "skip-verify", "preferred", "":
+		return nil
+	default:
+		return mysql.RegisterTLSConfig(ctls, &tls.Config{})
+	}
+}
+
 func urlToMySQLConfig(url string) (*mysql.Config, error) {
+	dsn := strings.Split(url, "?")
+	if len(dsn) > 1 {
+		dsnParams := dsn[1]
+		tlsConfig, err := extractDSNParam(dsnParams, "tls")
+		if err != nil {
+			return nil, errors.Wrap(err, "extracting `tls` DSN param")
```

Comment:
What do you think about returning a custom error type rather than a wrapped-error? Consumers could then check if the error returned is of a certain type, eg. `ErrExtractTLSParam`. Note that this tool could be used programmatically as well, so it could be useful to have a defined error. Of course, if/when we contribute this fix, I have no idea what the upstream maintainers might prefer. 🤷‍♂️ 

But using the custom error type would also eliminate the need to add a new import for `github.com/pkg/errors` and thus further reduce the drift from upstream.

Reply:
I think that would be ideal. For now this will unblock us though, but yea there are a lot of improvements we can make when we go to fix the upstream which is why I haven't opened a PR against the upstream yet, because there are definitely improvements we can make.